node/cancellation-v2.md

```diff
@@ -0,0 +1,343 @@
+  // Note that the Workflow is completely oblivious to cancellation in this example.
+  const result = await cancellationScope(async (scope) => {
+    scope.shield = true;
+    return httpGetJSON(url);
+  });
+  return result;
+}
+```
+
+#### `Context.cancelled` may be awaited upon in order to make Workflow aware of cancellation while waiting on shielded scopes
+
+```ts
+import { CancellationError, cancellationScope } from '@temporalio/workflow';
+import { httpGetJSON } from '@activities';
+
+export async function main(url: string) {
+  let result: any = undefined;
+  try {
+    const promise = cancellationScope({ shield: true }, () => httpGetJSON(url));
+    result = await Promise.race([Context.cancelled, promise]);
+  } catch (err) {
+    if (!(e instanceof CancellationError)) {
+      throw err;
+    }
+    // Prevent Workflow from completing so Activity can complete
+    result = await promise;
+  }
+  return result;
+}
+```
+
+#### Handle Workflow cancellation by an external client while an Activity is running
+
+```ts
+import { CancellationError, cancellationScope } from '@temporalio/workflow';
+import { httpPostJSON, cleanup } from '@activities';
+
+export async function main(url: string, data: any): Promise<void> {
+  try {
+    await httpPostJSON(url, data);
+  } catch (e) {
+    if (e instanceof CancellationError) {
+      console.log('Workflow cancelled');
+      // Cleanup logic goes in a shielded scope here
+      // If we'd run cleanup outside of a shielded scope it would've been cancelled
+      // before being started because the Workflow's root scope is cancelled.
+      await cancellationScope({ shield: true }, () => cleanup(url));
+    }
+    throw e; // <-- Fail the Workflow
+  }
+}
+```
+
+#### Complex flows may be achieved by nesting cancellation scopes
+
+```ts
+import { CancellationError, cancellationScope } from '@temporalio/workflow';
+import { setup, httpPostJSON, cleanup } from '@activities';
+
+export async function main(url: string) {
+  try {
+    return await cancellationScope(async (scope) => {
+      await cancellationScope({ shield: true }, () => setup());
+      scope.deadline = Date.now() + 1000; // 1 second after setup is complete
+      await httpPostJSON(url);
+    });
+  } catch (err) {
+    if (!(err instanceof CancellationError)) {
+      throw err;
+    }
+    await cancellationScope({ shield: true }, () => cleanup(url));
+  }
+}
+```
+
+#### Sharing promises between scopes
+
+```ts
+import { cancellationScope } from '@temporalio/workflow';
+import { activity1, activity2 } from '@activities';
+
+export async function main(): Promise<void> {
+  const p1 = activity1(); // <-- Start activity1 in root scope
+  const p2 = activity2(); // <-- Start activity2 in root scope
+
+  const scopePromise = cancellationScope(async (scope) => {
+    const first = await Promise.race([p1, p2]);
+    scope.cancel(); // Does not cancel activity1 or activity2 as they're tied to the root scope
+    return first;
```

Comment:
But `scopePromise` will be resolved as cancelled?

Reply:
No, `scopePromise` will be resolved with `first`.